fix: Warning elimination Wbad-function-cast

shikharvashistha opened the pull request:
Greetings of the day,

This pull adds the elimination of warning mentioned in #3255, kindly review the pull and let me know in case of changes if any.

Best Regards
@shikharvashistha

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

bazsi commented:
Ok to test.

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
> Build FAILURE

Hi @bazsi, I think the issue was with signing off the commit, as desired I've signed off.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented:
> > Build FAILURE
> 
> Hi @bazsi, I think the issue was with signing off the commit, as desired I've signed off.

This is commented by the end-to-end tests, which are unfortunately not publicly accessible.

I am approving the run of the GitHub based CI, please check if you change compiles cleanly with those jobs. If they are I can help with the end-to-end tests.

gaborznagy commented:
@kira-syslogng retest this please;

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

OverOrion commented:
Thank you for your contribution @shikharvashistha!

There are some problems with your commits, could you please sign-off all your commits?

```
[FAIL] 6094f164a fix: _calc_sec_between_time
       * This commit does not have a Signed-off-by line.
[FAIL] 5c7581a8a fix: typo
       * This commit does not have a Signed-off-by line.
```
---

Due to this modification `elapsed_time` is no longer a `gsize` variable, which is why the compiler (rightfully) gives a warning:

```
In file included from /__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c:31:
/__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c: In function ‘_calc_average’:
/__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c:277:52: error: conversion from ‘gdouble’ {aka ‘double’} to ‘long long int’ may change value [-Werror=float-conversion]
  277 |             evt_tag_long("delta_time_since_start", elapsed_time));
      |                                                    ^~~~~~~~~~~~
/__w/syslog-ng/syslog-ng/lib/messages.h:107:53: note: in definition of macro ‘msg_trace’
  107 |             msg_event_create(EVT_PRI_DEBUG, desc, ##tags, NULL ));          \
      |                                                     ^~~~
cc1: all warnings being treated as errors

```

You should cast it to `gsize` in the function call to be explicit about the type-conversion.

---
Also would you be so kind and fix the whitespace issue in `Makefile.am`?  It looks like `Wbad-function-cast` is not properly indented.

shikharvashistha commented:
> Thank you for your contribution @shikharvashistha!
> 
> There are some problems with your commits, could you please sign-off all your commits?
> 
> ```
> [FAIL] 6094f164a fix: _calc_sec_between_time
>        * This commit does not have a Signed-off-by line.
> [FAIL] 5c7581a8a fix: typo
>        * This commit does not have a Signed-off-by line.
> ```
> 
> Due to this modification `elapsed_time` is no longer a `gsize` variable, which is why the compiler (rightfully) gives a warning:
> 
> ```
> In file included from /__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c:31:
> /__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c: In function ‘_calc_average’:
> /__w/syslog-ng/syslog-ng/lib/stats/aggregator/stats-change-per-second.c:277:52: error: conversion from ‘gdouble’ {aka ‘double’} to ‘long long int’ may change value [-Werror=float-conversion]
>   277 |             evt_tag_long("delta_time_since_start", elapsed_time));
>       |                                                    ^~~~~~~~~~~~
> /__w/syslog-ng/syslog-ng/lib/messages.h:107:53: note: in definition of macro ‘msg_trace’
>   107 |             msg_event_create(EVT_PRI_DEBUG, desc, ##tags, NULL ));          \
>       |                                                     ^~~~
> cc1: all warnings being treated as errors
> ```
> 
> You should cast it to `gsize` in the function call to be explicit about the type-conversion.
> 
> Also would you be so kind and fix the whitespace issue in `Makefile.am`? It looks like `Wbad-function-cast` is not properly indented.

Greetings of the day,

kira-syslogng commented:
Build SUCCESS

OverOrion commented:
Thank you for addressing my notes so quickly!

Could you please squash the `fix: typo` and `Update Makefile.am` into the original `fix: Warning elimination Wbad-function-cast` commit? That would make the commit check failure go away (as `Update Makefile.am` doesn't conform [to our commit message format](https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#commits)).
If you need help with doing that please let me know.

----
Because you added the extra `-Wbad-function-cast` check the compiler now recognizes those as errors:

```__w/syslog-ng/syslog-ng/modules/basicfuncs/numeric-funcs.c:363:43: error: cast from function call of type ‘double’ to non-matching type ‘long int’ [-Werror=bad-function-cast]
  363 |   format_int64_padded(result, 0, ' ', 10, (gint64)ceil(number_as_double(n)));
      |                                           ^
/__w/syslog-ng/syslog-ng/modules/basicfuncs/numeric-funcs.c: In function ‘tf_num_floor’:
/__w/syslog-ng/syslog-ng/modules/basicfuncs/numeric-funcs.c:391:43: error: cast from function call of type ‘double’ to non-matching type ‘long int’ [-Werror=bad-function-cast]
  391 |   format_int64_padded(result, 0, ' ', 10, (gint64)floor(number_as_double(n)));
```
Could you please fix these as well?

If you think you have finished fixing it, then running `make -j $(nproc)` should compile syslog-ng successfully or the compiler would give you errors if there are any `bad-function-cast` errors left.

shikharvashistha commented:
> Thank you for addressing my notes so quickly!
> 
> Could you please squash the `fix: typo` and `Update Makefile.am` into the original `fix: Warning elimination Wbad-function-cast` commit? That would make the commit check failure go away (as `Update Makefile.am` doesn't conform [to our commit message format](https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#commits)). If you need help with doing that please let me know.
> 
> Because you added the extra `-Wbad-function-cast` check the compiler now recognizes those as errors:
> 
> ```
>   363 |   format_int64_padded(result, 0, ' ', 10, (gint64)ceil(number_as_double(n)));
>       |                                           ^
> /__w/syslog-ng/syslog-ng/modules/basicfuncs/numeric-funcs.c: In function ‘tf_num_floor’:
> /__w/syslog-ng/syslog-ng/modules/basicfuncs/numeric-funcs.c:391:43: error: cast from function call of type ‘double’ to non-matching type ‘long int’ [-Werror=bad-function-cast]
>   391 |   format_int64_padded(result, 0, ' ', 10, (gint64)floor(number_as_double(n)));
> ```
> 
> Could you please fix these as well?
> 
> If you think you have finished fixing it, then running `make -j $(nproc)` should compile syslog-ng successfully or the compiler would give you errors if there are any `bad-function-cast` errors left.

Hi, While running `make -j $(nproc)` I'm facing this issue, what could be the way out.
![image](https://user-images.githubusercontent.com/51105234/165685820-f504dbcf-380a-43b6-87aa-e706037749d2.png)

kira-syslogng commented:
Build FAILURE

OverOrion commented:
By default the Java module is enabled, if you do not have Java installed (or an incompatible version is installed) then I suggest you rerun the `configure` command like this `configure --disable-java` then run `make -j $(nproc)`

shikharvashistha commented:
> --disable-java



> By default the Java module is enabled, if you do not have Java installed (or an incompatible version is installed) then I suggest you rerun the `configure` command like this `configure --disable-java` then run `make -j $(nproc)`

But I think Java is installed on my system but it may be an incompatible version
![image](https://user-images.githubusercontent.com/51105234/165708824-54d7e0a7-395f-46a6-9168-5a321ae493e2.png)

shikharvashistha commented:
Made the changes and build locally
![image](https://user-images.githubusercontent.com/51105234/165710496-bd73d5cc-ab14-4d8a-b615-42067fe91eeb.png)

kira-syslogng commented:
Build SUCCESS

shikharvashistha commented:
Hi @OverOrion, why I'm unable to spot these errors locally when building

OverOrion commented:
Building with `cmake` yields a different result, I'll have to check why.

Currently building via `cmake` fails:
```
[ 99%] Building C object modules/diskq/tests/CMakeFiles/test_diskq_truncate.dir/test_diskq_truncate.c.o
/__w/syslog-ng/syslog-ng/modules/diskq/tests/test_diskq_truncate.c: In function ‘_calculate_full_disk_message_num’:
/__w/syslog-ng/syslog-ng/modules/diskq/tests/test_diskq_truncate.c:100:10: error: cast from function call of type ‘double’ to non-matching type ‘long int’ [-Werror=bad-function-cast]
  100 |   return (gint64) ceil((qdisk->options->disk_buf_size - QDISK_RESERVED_SPACE) / (double) msg_size);
      |          ^
[100%] Linking C executable test_reliable_backlog
cc1: all warnings being treated as errors
```

shikharvashistha commented:
> Building with `cmake` yields a different result, I'll have to check why.
> 
> Currently building via `cmake` fails:
> 
> ```
> [ 99%] Building C object modules/diskq/tests/CMakeFiles/test_diskq_truncate.dir/test_diskq_truncate.c.o
> /__w/syslog-ng/syslog-ng/modules/diskq/tests/test_diskq_truncate.c: In function ‘_calculate_full_disk_message_num’:
> /__w/syslog-ng/syslog-ng/modules/diskq/tests/test_diskq_truncate.c:100:10: error: cast from function call of type ‘double’ to non-matching type ‘long int’ [-Werror=bad-function-cast]
>   100 |   return (gint64) ceil((qdisk->options->disk_buf_size - QDISK_RESERVED_SPACE) / (double) msg_size);
>       |          ^
> [100%] Linking C executable test_reliable_backlog
> cc1: all warnings being treated as errors
> ```
Hi @OverOrion,

The Steps I used to build are as follows:

1.  Used autogen.sh
2.  ./configure --disable-java (after installing desired dependencies [flex & bison])
3.  Ran make -j $(nproc)

#### Output 
https://user-images.githubusercontent.com/51105234/165742944-efdb4010-c72e-4f24-9c77-a8a310a6de68.png

Kindly guide me if I'm doing something wrong

Thanks & Regards
@shikharvashistha

OverOrion commented:
Hi @shikharvashistha

By default cmake is configured with extra-warnings but this is not enabled for autotools.

You did nothing wrong (sorry, I did not check it thoroughly), here is a version which won't compile until you have fixed all the `bad-function-cast` errors:
1. `autogen.sh`
2. `./configure --enable-debug --enable-extra-warnings --disable-java CFLAGS="-Werror=bad-function-cast"`
3. `make -j $(nproc)`

This adds the extra `Werror=bad-function-cast` to `gcc` which will halt the compilation if it encounters this type of warning.

shikharvashistha commented:
> 

Hi @OverOrion, thanks for the prompt reply. I'll update the code accordingly.

Best
@shikharvashistha

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
Hi @OverOrion, I've fixed all the errors which came up during the above said procedure.

shikharvashistha commented:
Hi, @OverOrion. The procedure shows no issues to me locally (attaching image). Debugging the issue
![image](https://user-images.githubusercontent.com/51105234/166223902-567f726e-6e50-4758-b59d-f7f04d61a549.png)

OverOrion commented:
Hi @shikharvashistha,

You can run our unit tests with this: `make check`

Currently this is the output:
```
# TOTAL: 183
# PASS:  181
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: modules/diskq/tests/test_diskq
====================================

[----] ../modules/diskq/tests/test_diskq.c:558: Unexpected signal caught below this line!
[FAIL] diskq::test_diskq_statistics: CRASH!
Feed speed: 105236.57
[====] Synthesis: Tested: 9 | Passing: 8 | Failing: 1 | Crashing: 1 
FAIL modules/diskq/tests/test_diskq (exit status: 1)

FAIL: modules/diskq/tests/test_diskq_truncate
=============================================

Reliable disk-buffer state saved; filename='test_dq_no_truncate_wrap.rqf', qdisk_length='0'
Reliable disk-buffer state saved; filename='test_dq_truncate_on_write.rqf', qdisk_length='0'
Reliable disk-buffer state saved; filename='test_diskq_truncate_short_read.rqf', qdisk_length='0'
Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_0.rqf', qdisk_length='0'
Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_1.rqf', qdisk_length='0'
Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_default.rqf', qdisk_length='0'
Disk-buffer state saved; filename='test_dq_truncate1.qf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
[----] ../modules/diskq/tests/test_diskq_truncate.c:205: Unexpected signal caught below this line!
[FAIL] diskq_truncate::test_diskq_truncate_with_diskbuffer_used: CRASH!
Disk-buffer state saved; filename='test_dq_truncate2.qf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
Disk-buffer state saved; filename='test_dq_non_reliable_restart.rqf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
[----] ../modules/diskq/tests/test_diskq_truncate.c:217: Unexpected signal caught below this line!
[FAIL] diskq_truncate::test_diskq_truncate_without_diskbuffer_used: CRASH!
[----] ../modules/diskq/tests/test_diskq_truncate.c:546: Unexpected signal caught below this line!
[FAIL] diskq_truncate::test_non_reliable_diskq_restart_with_truncation_disabled: CRASH!
[====] Synthesis: Tested: 9 | Passing: 6 | Failing: 3 | Crashing: 3 
FAIL modules/diskq/tests/test_diskq_truncate (exit status: 1)
```

I can see that you've changed
 `
      POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qoverflow), &path_options);
`
to 
`      POINTER_TO_LOG_PATH_OPTIONS(msg, &path_options);`

Which are not equivalent,  `g_queue_pop_head` removes the first element of the queue, which is a sideffect that needs to be reserved.

shikharvashistha commented:
> Hi @shikharvashistha,
> 
> You can run our unit tests with this: `make check`
> 
> Currently this is the output:
> 
> ```
> # TOTAL: 183
> # PASS:  181
> # SKIP:  0
> # XFAIL: 0
> # FAIL:  2
> # XPASS: 0
> # ERROR: 0
> 
> .. contents:: :depth: 2
> 
> FAIL: modules/diskq/tests/test_diskq
> ====================================
> 
> [----] ../modules/diskq/tests/test_diskq.c:558: Unexpected signal caught below this line!
> [FAIL] diskq::test_diskq_statistics: CRASH!
> Feed speed: 105236.57
> [====] Synthesis: Tested: 9 | Passing: 8 | Failing: 1 | Crashing: 1 
> FAIL modules/diskq/tests/test_diskq (exit status: 1)
> 
> FAIL: modules/diskq/tests/test_diskq_truncate
> =============================================
> 
> Reliable disk-buffer state saved; filename='test_dq_no_truncate_wrap.rqf', qdisk_length='0'
> Reliable disk-buffer state saved; filename='test_dq_truncate_on_write.rqf', qdisk_length='0'
> Reliable disk-buffer state saved; filename='test_diskq_truncate_short_read.rqf', qdisk_length='0'
> Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_0.rqf', qdisk_length='0'
> Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_1.rqf', qdisk_length='0'
> Reliable disk-buffer state saved; filename='test_dq_truncate_size_ratio_default.rqf', qdisk_length='0'
> Disk-buffer state saved; filename='test_dq_truncate1.qf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
> [----] ../modules/diskq/tests/test_diskq_truncate.c:205: Unexpected signal caught below this line!
> [FAIL] diskq_truncate::test_diskq_truncate_with_diskbuffer_used: CRASH!
> Disk-buffer state saved; filename='test_dq_truncate2.qf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
> Disk-buffer state saved; filename='test_dq_non_reliable_restart.rqf', qout_length='0', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
> [----] ../modules/diskq/tests/test_diskq_truncate.c:217: Unexpected signal caught below this line!
> [FAIL] diskq_truncate::test_diskq_truncate_without_diskbuffer_used: CRASH!
> [----] ../modules/diskq/tests/test_diskq_truncate.c:546: Unexpected signal caught below this line!
> [FAIL] diskq_truncate::test_non_reliable_diskq_restart_with_truncation_disabled: CRASH!
> [====] Synthesis: Tested: 9 | Passing: 6 | Failing: 3 | Crashing: 3 
> FAIL modules/diskq/tests/test_diskq_truncate (exit status: 1)
> ```
> 
> I can see that you've changed `POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qoverflow), &path_options);` to ` POINTER_TO_LOG_PATH_OPTIONS(msg, &path_options);`
> 
> Which are not equivalent, `g_queue_pop_head` removes the first element of the queue, which is a sideffect that needs to be reserved.

Hi, @OverOrion. Thanks will update as desired

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
Hi @OverOrion `make check` shows unit tests are disabled.

Kokan commented:
The unit tests requires Criterion to be installed: see https://github.com/Snaipe/Criterion

shikharvashistha commented:
> The unit tests requires Criterion to be installed: see https://github.com/Snaipe/Criterion

Hi, @Kokan. Thanks for the guidance. I've updated the code as desired.

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
> Build FAILURE
> 

I've disabled python too like I did java. It still shows Python.h as no include directory, what could be the issue.

shikharvashistha commented:
> Hi @shikharvashistha,
> 
> I've quickly checked the current status of this PR and here are my notes:
> 
> 1. Please rebase against current master as there had been a test fix ([diskq: refactor test_diskq_statistics #4004](https://github.com/syslog-ng/syslog-ng/pull/4004))
> 2. Please do not _remove_ the `POINTER_TO_LOG_PATH_OPTIONS` macro because by manually expanding it someone might inadvertently forgot to make changes in all those blocks / it helped reduce code duplication / etc. Was there a reason for it's removal?
> 3. Note that removing `g_queue_pop_head` in `modules/diskq/qdisk.c:980` (and other places) will break the test, as they rely on that side-effect, please keep those side effects while fixing `Wbad-function-cast` errors because they are needed.
> 
> > I've disabled python too like I did java. It still shows Python.h as no include directory, what could be the issue.
> 
> Could you please clarify what you mean by it?

Hi, @OverOrion, 

1. Rebased ✅ 
2. The reason I removed `POINTER_TO_LOG_PATH_OPTIONS` macro as `(GPOINTER_TO_INT(ptr)` is causing a `-Wbad-function-cast` error.
3.  Fixed 👍 

> Could you please clarify what you mean by it?
- Configure command I'm using is `./configure --enable-debug --enable-extra-warnings --disable-java-modules --with-python=3 --disable-java CFLAGS="-Werror=bad-function-cast"`
- While performing tests facing this issue

![image](https://user-images.githubusercontent.com/51105234/167848619-86b16272-ff5a-4c68-9ab5-5cebbd6781cf.png)

kira-syslogng commented:
Build FAILURE

OverOrion commented:
Hi @shikharvashistha,

1. Currently your branch is based on 7b41872d7, which is not what I mentioned. Please make sure your fork is up to date with upstream master (click on "Fetch and merge"), then rebase against it
![image](https://user-images.githubusercontent.com/15684829/168006712-a363daa8-356d-4a4a-82c9-fbfdcda4e074.png)
Your PR should be based on on the latest master, which is currently 209b1cd4af5f9f68a05d03f62d415ecb4bb68b41
2. > The reason I removed POINTER_TO_LOG_PATH_OPTIONS macro as (GPOINTER_TO_INT(ptr) in the macro is causing a -Wbad-function-cast error.

Not exactly, see this example:
```
#define GPOINTER_TO_INT(p)	((gint)   (p))
#define POINTER_TO_LOG_PATH_OPTIONS(ptr, lpo) (lpo)->ack_needed = (GPOINTER_TO_INT(ptr) & ~0x80000000)


static inline LogMessage *
_pop_head_qout(LogQueueDiskNonReliable *self, LogPathOptions *path_options)
{
  LogMessage *msg = g_queue_pop_head(self->qout);
  POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qout), path_options);
  log_queue_memory_usage_sub(&self->super.super, log_msg_get_size(msg));

  return msg;
}
```

`  POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qout), path_options);` will be expanded to this:
``` 
  (lpo)path_options->ack_needed = (GPOINTER_TO_INT(g_queue_pop_head(self->qout)) & ~0x80000000) 
// and finally to this:
  (lpo)path_options->ack_needed = (((gint)   (g_queue_pop_head(self->qout)))) & ~0x80000000)
```

The reason you got `Wbad-function-cast error` is because casting the return value of `g_queue_pop_head` (which returns a `gpointer`) to `gint`. Saving the return value of the function to an intermediate variable should be the way to go.

Other notes:
- please do not leave commented out codes in your commits
- make sure that you ran `make style-format` before pushing your commits

Regarding the missing python header:
It seems that you have not installed the needed python development packages, on Debian based systems this should take care of it:
`sudo apt-get install python3-dev`

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
Hi, @OverOrion. I've made the suggested changes and now only one test fails which is (attaching ss). What could be the issue as it's not logged.

![image](https://user-images.githubusercontent.com/51105234/168039216-d2ec47e6-9306-4bdf-9817-269d581b2e02.png)

kira-syslogng commented:
Build SUCCESS

MrAnno commented on Makefile.am:
```diff
@@ -88,7 +88,8 @@ AM_CFLAGS += \
 	-Woverride-init \
-	-Wfloat-conversion
+	-Wfloat-conversion \
+        -Wbad-function-cast
```
Please use tabs instead of spaces in Makefiles.

shikharvashistha replied:
Fixed 👍

OverOrion replied:
The first line is indented with 4 spaces, the second line is indented with 8 spaces.
Make sure that you use tab characters as your indentation as some editors may insert spaces instead of tabs depending on their configuration.

OverOrion replied:
The first line is indented with 4 spaces, the second line is indented with 8 spaces.
Make sure that you use tab characters as your indentation as some editors may insert spaces instead of tabs depending on their configuration.

shikharvashistha replied:
Yeah true it was the editor configuration issue

MrAnno commented on modules/afsocket/afsocket-source.c:
```diff
@@ -847,8 +847,9 @@ _sd_open_stream(AFSocketSourceDriver *self)
+      gpointer fd = cfg_persist_config_fetch(cfg, afsocket_sd_format_listener_name(self));
       sock = GPOINTER_TO_UINT(
-               cfg_persist_config_fetch(cfg, afsocket_sd_format_listener_name(self))) -
+               fd) -
```
This could be merged into a single line now that you've extracted the config_persist call.

`sock = GPOINTER_TO_UINT(fd) - 1;`

shikharvashistha replied:
Fixed 👍

MrAnno commented on modules/diskq/logqueue-disk-non-reliable.c:
```diff
@@ -223,7 +225,9 @@ _ack_backlog(LogQueue *s, gint num_msg_to_ack)
-      POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(self->qbacklog), &path_options);
+      next_msg = g_queue_pop_head(self->qbacklog);
+      gint size = GPOINTER_TO_INT(next_msg);
+      POINTER_TO_LOG_PATH_OPTIONS(size, &path_options);
```
This warning type is tricky as it warns on casts, which are usually deliberate decisions.
`POINTER_TO_LOG_PATH_OPTIONS()` contains such an intentional cast, hence the problem.

`Wbad-function-cast` can be useful for detecting changed interfaces at API separation points, but in this special case, it just causes problems.

Removing this macro is not a good idea as it makes the code hard to read.
Maybe a reimplementation could help, for example:
```
#define POINTER_TO_LOG_PATH_OPTIONS(ptr, lpo) ({ \
  gpointer p = ptr;
  LogPathOptions *opts = lpo;
  (lpo)->ack_needed = (GPOINTER_TO_INT(p) & ~0x80000000) \
})
```

Or even better: a real C function could work too.

shikharvashistha replied:
Fixed as desired 👍

kira-syslogng commented:
Build FAILURE

shikharvashistha commented:
Greetings of the day Mr. @MrAnno,

As suggested by Mr. @bazsi that end-to-end tests are not publicly available. Can you kindly share the Jenkins logs for Kira build, so that I can take the task towards completion.

Thanks & Regards
@shikharvashistha

MrAnno commented:
@shikharvashistha Kira shows a lot of end-to-end test failures related to the disk-buffer module.
The output is huge, so I think it's not worth sharing it.

The root cause of the problem is the non-equivalent transformations in the disk-buffer module.
It's basically about my previous review comment: instead of removing `POINTER_TO_LOG_PATH_OPTIONS()`, you can use a function or the suggested macro modification at every occurrence.

kira-syslogng commented:
Build SUCCESS

OverOrion commented on modules/diskq/logqueue-disk-non-reliable.c:
```diff
@@ -36,6 +36,10 @@ typedef struct
 
+
+/* helper function for _ack_backlog() */
+void _ack_backlog_one_message(LogQueueDiskNonReliable *self, gpointer data, LogPathOptions *path_options);
```
Function declarations should be in header files, but from what I can see this function is not used anywhere. How about removing it?

shikharvashistha replied:
Hi, @OverOrion. Yeah I've implemented that to make the code cleaner (to have a function instead of macro revamp), but that would have required to change implementation at all the places, So instead to avoid code duplication I've revamped the macro as suggested my @MrAnno but forgot to remove the same declaration. Removed it 👍

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

OverOrion commented:
Hi @shikharvashistha ,

Thank you for addressing the review notes, the changes themselves look good, but would you be so kind and squash them into 1-2 commits?
There is a merge commit which is not needed either.

shikharvashistha commented on modules/diskq/qdisk.h:
```diff
@@ -35,7 +35,7 @@
  * which has side effects.
  */
-#define POINTER_TO_LOG_PATH_OPTIONS(ptr, lpo) (lpo)->ack_needed = (GPOINTER_TO_INT(ptr) & ~0x80000000)
+#define POINTER_TO_LOG_PATH_OPTIONS(ptr, lpo) ({gpointer p = ptr; LogPathOptions *opts = lpo; (lpo)->ack_needed = (GPOINTER_TO_INT(p) & ~0x80000000); opts;})
```
Fixed 👍

kira-syslogng commented:
Build SUCCESS

MrAnno commented:
You can set your preferred editor with `git config`, for example:
```
git config --global core.editor vim
```

shikharvashistha commented:
> You can set your preferred editor with `git config`, for example:
> 
> ```
> git config --global core.editor vim
> ```

Tried not working for me 😓

kira-syslogng commented:
Build SUCCESS

shikharvashistha commented:
Hi @MrAnno As I was facing issues while rebasing withing my editor/terminal therefore I've created a new pull (#4009) for the same.
